### ISSUE TITLE
Allow minimal distance-based edges in a habitat graph

### ISSUE THREAD

**Issue by rdnfn** [ai4er-cdt/gtc-biodiversity#22]
**Description:** Currently graph edges are only added to the graph when two polygons overlap or touch. For habitats there will almost always be a certain distance a species is able to travel across land that is not part of the habitat. Therefore we want to be able to add edges between polygons if they are within a certain _maximum travel distance_.

**Potential solution:** One way would be to add an edge to all polygons within a certain range (e.g. 10m) in the initial creation of the (parent-)graph and add an edge-attribute of `min_distance` between two polygons. Then when we add habitat subgraphs later we could easily filter the edges such that their `min_distance` is withing the `max_travel_distance`.

**Further notes:**
The `create_habitat` method of the `GeoGraph` class might take the following arguments

- `name` (e.g. `"eagle habitat"`)
- list of landcover classes that are part of the habitat (e.g. [`"pine forests", ...]`)
- `max_travel_distance`: maximum travel distance outside habitat (e.g. 50m)
- list of landcover classes that are uncrossable/impassable (e.g.` ["motorway",...]`)

**Comment by herbiebradley** [comment]
The RTree implementation returns a list of polygons which are within the rectangular bounding box of the passed `Geometry`. Therefore the best algorithm for this may be:

```
for each polygon:
    define a rectangular polygon object from the bounding box of the original polygon by adding on the max_travel_distance
    query the rtree with this object to get a list of candidate polygons
    for each candidate polygon:
        check if the min_distance between the candidate and the original polygon is within max_travel_distance
        if yes, add an edge with min_distance
```

Note that we don't have to do this on graph creation, but we can do it at any time as long as we have the RTree and a way to iterate over polygons/nodes (maybe we should add the RTree as a class attribute). After looking into the `networkx` docs, I think we will actually need to create a separate graph object for each habitat (correct me if I'm wrong), so the `create_habitat` function might duplicate the graph, add these edges in, then add the graph object for that habitat into a class attribute dictionary.

**Comment by rdnfn** [comment]
That pseudocode looks good! The idea with doing adding the `min_distance` attribute at graph creation was to reuse the `min_distance` computations (because I think those will be the most computationally expensive part, although I haven't tested this). I agree that doing all the computations at the graph creation stage might be limiting (as it would set a max value for `max_travel_distance` for all subsequent habitats), maybe there is a way we can save the computations just between habitat creations? (so that when we create `Habitat B` after `Habitat A`, all `min_distance` computations from `Habitat A` can be reused. I might also be overthinking it though because `min_distance` not that computationally expensive after all.)

**Comment by herbiebradley** [comment]
I expect that calculating the distance will be the most expensive part of this - so the best caching solution I can see would be to store all `min_distance`'s calculated as edge attributes, then when we create `Habitat B` we can reuse the ones that we have and calculate ones that we don't.

**Comment by herbiebradley** [comment]
Here's an alternative method based on what we know now.
```
for each polygon:
    define a new_polygon by using buffer(max_travel_distance)
    query the rtree with this object to get a list of candidate polygons
    for each candidate polygon:
        check if the new_polygon intersects with the candidate polygon
        if yes, add an edge (maybe add min_distance as an edge attribute)
```
This should be significantly faster since we only ever calculate the distance between nodes that are actually part of the habitat.

**Comment by herbiebradley** [comment]
Closed by #31
